### ISSUE TITLE
cache warmup Neo4j 3.2 

### ISSUE THREAD

**Issue by zunio** [neo4j/neo4j#9597]
Hi,

I can't able to cache all the graph in memory, with last Neo4j version

Neo4J Version: 3.2.1
OS: CentOS 6.7
Size Graph: 19g
dbms.memory.pagecache.size = 20g
Ram available= 30.5 g

[export_apoc.xlsx](https://github.com/neo4j/neo4j/files/1112455/export_apoc.xlsx)

[neo4j_conf.docx](https://github.com/neo4j/neo4j/files/1112494/neo4j_conf.docx)

I have tried both the query both apoc approach that are described here 
https://neo4j.com/developer/kb/warm-the-cache-to-improve-performance-from-cold-start/

but the Neo4j's process used only the 40% of memory.

With 3.0.1 Version with the same configuration we don't have this issue.

You can find in the attachment files:
- the neo4j.conf file
- the output of apoc.warmup.run

**Comment by chrisvest** [comment]
How much of that 19 GiB of data is in the `index/`, `schema/` directories, and `*transaction.db*` files?

Indexes, schema or manual, are not cached in the same way, and transaction log files are not cached at all because they are mostly only written to.

**Comment by zunio** [comment]
`du -hc *store.db* ` -> 19GB

schema/ -> 2.4 GB
index/ -> 4K
*transaction.db* -> empty

**Comment by jexp** [comment]
Apoc's warm up by default doesn't load property pages.

You can run `call apoc.warmup.run(true)` to also load properties.

**Comment by jexp** [comment]
When did you observe the fuller usage of the machine memory with 3.0.x ? 
After warmup or after running production workload?

**Comment by zunio** [comment]
We have a warmup script . When we start neo4j, we also launch 
`MATCH (n)
OPTIONAL MATCH (n)-[r]->()
RETURN count(n.prop) + count(r.prop);`

When the script end, we observe  memory usage.

**Comment by zunio** [comment]
I've tryed on a new and clean ec2 instance but I've the same problem.

It looks like that the config parameter dbms.memory.pagecache.size doesn't being read.
If I comment it, the process use the same amount of memory (40%) than I set the parameter.

It's quite strange....

**Comment by GiuliaBusnelli** [comment]
Hi @chrisvest, @jexp
I am a colleague of @zunio and I am wondering if you can help us on our issue with cache warm-up.
I collected some other data, here they are:

### Neo4j v3.0.1, with graph imported through BatchInserter Java API v3.0.1

**Graph dimensions:**

- total: 19G
- /index folder: 4K
- /schema folder: 2.6G
- `*store.db*`: 17G of which 11G for neostore.propertystore.db file
- `*transaction.db*`: 4K

**Memory usage by Neo4j 3.0.1 process after cache warm-up:**

- VIRT: 27G
- RES: 21G

### Neo4j v3.2.1, with graph imported through BatchInserter Java API v3.2.1

**Graph dimensions:**

- total: 9.6G
- /index folder: 4K
- /schema folder: 1.3G
- `*store.db*`: 8.4G of which 5.3G for neostore.propertystore.db file
- `*transaction.db`*: 4K

**Memory usage by Neo4j 3.2.1 process before cache warm-up:**

- VIRT: 9G
- RES: 0.8G

**Memory usage by Neo4j 3.2.1 process after cache warm**-up:

- VIRT: 9.6G
- RES: 3.7G

In both cases, the page cache configuration is set to 20g and our warm-up query is the following:
`MATCH (n) OPTIONAL MATCH (n)-[r]->() RETURN count(n.prop) + count(r.prop)`

What do you think about those numbers?
Thank you very much.
Best, Giulia

**Comment by jexp** [comment]
That doesn't load all the properties, e.g. heavy properties are only loaded on access.
Other properties are only loaded if their block is loaded, which doesn't happy if the requested property is found earlier.

**Comment by zunio** [comment]
Hi @jexp ,
is there a way to force the loading of all properties?

**Comment by jexp** [comment]
Depends for non-long strings / arrays you can use `call apoc.warmup.run(true)` which runs also in parallel.

**Comment by zunio** [comment]
we've tried also the apoc but the problem remains

**Comment by chrisvest** [comment]
After these warmups, do you see read IO when running queries, and if so, to what files?

**Comment by zunio** [comment]
No, I can't see any IO.

I checked with iotop (-p).

On the other hand, if I execute the pageRankWithCypher (with write: true )all the graph is cached.

**Comment by chrisvest** [comment]
@zunio We have automatic page cache reheating as part of 3.4: https://neo4j.com/docs/operations-manual/current/performance/disks-ram-and-other-tips/#performance-page-cache-warmup

I'll close this, but feel free to reopen or create another issue if you run into any problems with this.
